dmriprep/workflows/fmap/base.py

```diff
@@ -0,0 +1,141 @@
+
+        wf.connect([
+            (inputnode, sdc_estimate_wf, [("dwi_reference", "inputnode.in_data")]),
+        ])
+        if generate_report:
+            from sdcflows.interfaces.reportlets import FieldmapReportlet
+            pepolar_report = pe.Node(FieldmapReportlet(reference_label="SDC'd B0"),
+                                     name="pepolar_report")
+            ds_report_pepolar = pe.Node(DerivativesDataSink(
+                base_directory=str(config.execution.work_dir / "reportlets"), keep_dtype=True,
+                desc="pepolar"), name="ds_report_pepolar")
+            ds_report_pepolar.inputs.source_file = epi_targets[0]
+            wf.connect([
+                (sdc_estimate_wf, pepolar_report, [
+                    ("outputnode.fieldmap", "fieldmap"),
+                    ("outputnode.corrected", "reference"),
+                    ("outputnode.corrected_mask", "mask")]),
+                (pepolar_report, ds_report_pepolar, [("out_report", "in_file")]),
+            ])
+
+    return wf
+
+
+def init_pepolar_estimate_wf(debug=False, generate_report=True, name="pepolar_estimate_wf"):
+    """Initialize a barebones TOPUP implementation."""
+    from nipype.interfaces.afni import Automask
+    from nipype.interfaces.fsl.epi import TOPUP
+    from niworkflows.interfaces.nibabel import MergeSeries
+    from ...interfaces.images import RescaleB0
+    wf = Workflow(name=name)
+
+    inputnode = pe.Node(niu.IdentityInterface(fields=["metadata", "in_data"]),
+                        name="inputnode")
+    outputnode = pe.Node(niu.IdentityInterface(fields=["fieldmap", "corrected", "corrected_mask"]),
+                         name="outputnode")
+
+    concat_blips = pe.Node(MergeSeries(), name="concat_blips")
+
+    topup = pe.Node(TOPUP(config=_pkg_fname(
+        "dmriprep", f"data/flirtsch/b02b0{'_quick' * debug}.cnf")), name="topup")
+
+    pre_mask = pe.Node(Automask(dilate=1, outputtype="NIFTI_GZ"),
+                       name="pre_mask")
+    rescale_corrected = pe.Node(RescaleB0(), name="rescale_corrected")
+    post_mask = pe.Node(Automask(outputtype="NIFTI_GZ"),
+                        name="post_mask")
+    wf.connect([
+        (inputnode, concat_blips, [("in_data", "in_files")]),
+        (inputnode, topup, [(("metadata", _get_ro), "readout_times"),
+                            (("metadata", _get_pedir), "encoding_direction")]),
+        (concat_blips, topup, [("out_file", "in_file")]),
+        (topup, pre_mask, [("out_corrected", "in_file")]),
+        (pre_mask, rescale_corrected, [("out_file", "mask_file")]),
+        (topup, rescale_corrected, [("out_corrected", "in_file")]),
+        (topup, outputnode, [("out_field", "fieldmap")]),
+        (rescale_corrected, post_mask, [("out_ref", "in_file")]),
+        (rescale_corrected, outputnode, [("out_ref", "corrected")]),
+        (post_mask, outputnode, [("out_file", "corrected_mask")]),
+    ])
+
+    return wf
+
+
+def _get_ro(metadata):
```

Comment:
Might be worth adding a fallback function for calculating the readout from other metadata available (if this isn't already covered elsewhere?)
```
def compute_readout(params):
    epi_factor = 1.0
    acc_factor = 1.0
    try:
        if params['epi_factor'] > 1:
            epi_factor = float(params['epi_factor'] - 1)
    except:
        pass
    try:
        if params['acc_factor'] > 1:
            acc_factor = 1.0 / params['acc_factor']
    except:
        pass
    return acc_factor * epi_factor * params['echospacing']
```

Reply:
Yup, that should use this https://github.com/nipreps/sdcflows/blob/28b084ff0cde85fbc34c1eb354401751e563aec8/sdcflows/interfaces/fmap.py#L465-L532

Reply:
BTW, the -1 should be subtracted from the effective epi factor, not the actual factor. :)

Reply:
Awesome.